// can-simple-dom/dist/amd/dom.js
/*can-simple-dom@0.2.7#dom*/
define([
    'exports',
    './simple-dom/document/node',
    './simple-dom/document/element',
    './simple-dom/document',
    './simple-dom/html-parser',
    './simple-dom/html-serializer',
    './simple-dom/void-map'
], function (exports, _simpleDomDocumentNode, _simpleDomDocumentElement, _simpleDomDocument, _simpleDomHtmlParser, _simpleDomHtmlSerializer, _simpleDomVoidMap) {
    'use strict';
    var _interopRequire = function (obj) {
        return obj && obj.__esModule ? obj['default'] : obj;
    };
    var Node = _interopRequire(_simpleDomDocumentNode);
    var Element = _interopRequire(_simpleDomDocumentElement);
    var Document = _interopRequire(_simpleDomDocument);
    var HTMLParser = _interopRequire(_simpleDomHtmlParser);
    var HTMLSerializer = _interopRequire(_simpleDomHtmlSerializer);
    var voidMap = _interopRequire(_simpleDomVoidMap);
    exports.Node = Node;
    exports.Element = Element;
    exports.Document = Document;
    exports.HTMLParser = HTMLParser;
    exports.HTMLSerializer = HTMLSerializer;
    exports.voidMap = voidMap;
    Object.defineProperty(exports, '__esModule', { value: true });
});
