// testee-client/lib/adapters/jasmine.js
var _ = require('underscore');
var guid = require('./../guid');

var TesteeReporter = function (runner) {
    this.runner = runner;
    this.currentSuite = null;
    this.currentSpec = null;
    this.suites = {};
    this.specs = {};
    this.failed = 0;
    this.total = 0;
};

var fakeFocusedSuite = {
    id: 'focused',
    description: 'focused specs',
    fullName: 'focused specs'
};

var __suites = {}, __specs = {};
function getSuite(suite) {
    __suites[suite.id] = _.extend(__suites[suite.id] || {}, suite);
    return __suites[suite.id];
}
function getSpec(spec) {
    __specs[spec.id] = _.extend(__specs[spec.id] || {}, spec);
    return __specs[spec.id];
}

_.extend(TesteeReporter.prototype, {
    log: _.noop,

    id: function(obj, id) {
        if(obj[id]) {
            return obj[id];
        }

        return (obj[id] = guid());
    },

    jasmineStarted: function () {
        var id = this.runId = guid();
        this.runner.start({
            id: id,
            environment: navigator.userAgent,
            runner: 'Jasmine',
            time: new Date().getTime()
        });
    },

    jasmineDone: function () {
        if (this.currentSuite) {
            this.suiteDone(fakeFocusedSuite);
        }
        this.runner.end({
            id: this.runId,
            failed: this.failed,
            total: this.total,
            passed: this.total - this.failed
        });
    },

    suiteStarted: function (suite) {
        suite = getSuite(suite);
        suite._parent = this.currentSuite;
        this.currentSuite = suite;
        if (suite._parent !== null) {
            if (!suite._parent.started) {
                this.suiteStarted(suite._parent);
            }
        }

        if (suite._parent !== null) {
            this.runner.suite({
                title: suite.description,
                parent: this.id(this.suites, suite._parent.id),
                id: this.id(this.suites, suite.id)
            });
        } else {
            this.runner.suite({
                title: suite.description,
                root: true,
                id: this.id(this.suites, suite.id),
                parent: this.runId
            });
        }

        suite.started = true;
    },

    suiteDone: function (suite) {
        suite = getSuite(suite);
        this.currentSuite = suite._parent;
        if (suite.started) {
            this.runner.suiteEnd({
                id: this.id(this.suites, suite.id)
            });
        }
    },

    specStarted: function (spec) {
        if(!this.currentSuite) {
            this.suiteStarted(fakeFocusedSuite);
        }
        spec = getSpec(spec);
        spec._suite = this.currentSuite;
        if (!spec._suite.started) {
            this.suiteStarted(spec._suite);
        }
        spec.startTime = new Date();
        if (!spec._suite.startTime) {
            spec._suite.startTime = spec.startTime;
        }
        this.currentSpec = spec;
        this.runner.test({
            title: spec.description,
            parent: this.id(this.suites, spec._suite.id),
            id: this.id(this.specs, spec.id)
        });
    },

    specDone: function (result) {
        if (result.failedExpectations.length > 0) {
            var message = result.failedExpectations[0].message;
            var stack = result.failedExpectations[0].stack;

            this.failed++;
            this.total++;

            this.runner.fail({
                id: this.id(this.specs, this.currentSpec.id),
                err: {
                    message: message,
                    stack: stack
                }
            });
        } else if (result.passedExpectations.length > 0) {
            var duration = (new Date().getTime() - this.currentSpec.startTime) || 0;

            this.total++;

            this.runner.pass({
                duration: duration,
                id: this.id(this.specs, this.currentSpec.id)
            });
        }
    }
});

module.exports = function (jasmine, runner) {
    jasmine.getEnv().addReporter(new TesteeReporter(runner));
};

module.exports.Reporter = TesteeReporter;
