// testee-client/lib/adapters/jasmine-legacy.js
var _ = require('underscore');
var guid = require('./../guid');

var TesteeReporter = function (runner) {
	this.runner = runner;
	this.suites = {};
	this.specs = {};
	this.failed = 0;
	this.total = 0;
};

_.extend(TesteeReporter.prototype, {
	log: _.noop,

	id: function(obj, id) {
		if(obj[id]) {
			return obj[id];
		}

		return (obj[id] = guid());
	},

	reportRunnerStarting: function () {
		var id = this.runId = guid();
		this.runner.start({
			id: id,
			environment: navigator.userAgent,
			runner: 'Jasmine',
			time: new Date().getTime()
		});
	},

	reportRunnerResults: function () {
		this.runner.end({
			id: this.runId,
			failed: this.failed,
			total: this.total,
			passed: this.total - this.failed
		});
	},

	reportSpecResults: function (spec) {
		if (spec.results_.failedCount) {
			var message = spec.results_.items_[0].message;
			var stack = spec.results_.items_[0].trace.stack;

			this.failed++;
			this.total++;

			this.runner.fail({
				id: this.id(this.specs, spec.id),
				err: {
					message: message,
					stack: stack
				}
			});
		} else if (spec.results_.passedCount) {
			var duration = (new Date().getTime() - spec.startTime) || 0;

			this.total++;

			this.runner.pass({
				duration: duration,
				id: this.id(this.specs, spec.id)
			});
		}
	},

	startSuite: function (suite) {
		if (suite.parentSuite !== null) {
			if (!suite.parentSuite.started) {
				this.startSuite(suite.parentSuite);
			}
		}

		if (suite.parentSuite !== null) {
			this.runner.suite({
				title: suite.description,
				parent: this.id(this.suites, suite.parentSuite.id),
				id: this.id(this.suites, suite.id)
			});
		} else {
			this.runner.suite({
				title: suite.description,
				root: true,
				id: this.id(this.suites, suite.id),
        parent: this.runId
			});
		}

		suite.started = true;
	},

	reportSpecStarting: function (spec) {
		if (!spec.suite.started) {
			this.startSuite(spec.suite);
		}
		spec.startTime = new Date();
		if (!spec.suite.startTime) {
			spec.suite.startTime = spec.startTime;
		}
		this.runner.test({
			title: spec.description,
			parent: this.id(this.suites, spec.suite.id),
			id: this.id(this.specs, spec.id)
		});
	},

	reportSuiteResults: function (suite) {
		if (suite.started) {
			this.runner.suiteEnd({
				id: this.id(this.suites, suite.id)
			});
		}
	}
});

module.exports = function (jasmine, runner) {
	jasmine.getEnv().addReporter(new TesteeReporter(runner));
};

module.exports.Reporter = TesteeReporter;
